[JENKINS-46247] Fix credentials scope in withAWS step and add a crede…

Dohbedoh opened the pull request:
…ntials dropdown

@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

Dohbedoh commented:
Tested the Descriptor locally:

![listboxmodel](https://user-images.githubusercontent.com/6858329/29907478-e8ab043c-8e5e-11e7-81bc-3f88c8425455.png)

hoegertn commented:
Looks good. I will test it.

There are some commented lines. Do you want to remove them before merging?
Should I wait for another reviewer?

Dohbedoh commented:
Thanks. Let me remove the comments.

> Should I wait for another reviewer?

With Jenkins World on, not sure if we will get a review this week. The build is failing so it may be best ? The tests pass locally.

hoegertn commented:
The failing build seems to be a problem with TravisCI. Maybe I will try other build engines or disable them completely.

Dohbedoh commented:
Got it. 
I cleaned up the comments. I also noticed I was using methods from previous versions of the credentials API. I switched to the recommended one.

Dohbedoh commented:
@hoegertn Note that you can simply add a Jenkinsfile with `buildPlugin()`and use https://ci.jenkins.io/

hoegertn commented:
Sounds great. Will give it a try.

hoegertn commented:
Please rebase onto master, as the build is fixed.

Dohbedoh commented on src/main/java/de/taimos/pipeline/aws/WithAWSStep.java:
```diff
@@ -176,12 +198,11 @@ public void expand(@Nonnull EnvVars envVars) throws IOException, InterruptedExce
-				UsernamePasswordCredentials usernamePasswordCredentials = CredentialsMatchers.firstOrNull(credentials, matcher);
+				StandardUsernamePasswordCredentials usernamePasswordCredentials = CredentialsProvider.findCredentialById(this.step.getCredentials(),
+						StandardUsernamePasswordCredentials.class, run, Collections.<DomainRequirement>emptyList());
 				if (usernamePasswordCredentials != null) {
```
Isn't the `RuntimeException` enough here ? Do you mean a warning in the console output ?

cyrille-leclerc replied:
my mistake,, I didn't see the "else" block, I didn't expect a "double negation". The Runtime Exception is indeed a clear message.

Dohbedoh commented on src/main/java/de/taimos/pipeline/aws/WithAWSStep.java:
```diff
@@ -123,47 +127,65 @@ public String getExternalId() {
+                                    ? Tasks.getAuthenticationOf((Queue.Task) context)
+                                    : ACL.SYSTEM,
+						    context,
+                            StandardUsernamePasswordCredentials.class,
```
The only cons I can think of is that `AmazonWebServicesCredentials` carries a redundant information (the role). The `withAWS` step has a parameter for this already. So it could be confusing or it could be integrated accordingly with one taking precedence (the `role` specified in the `withAWS` step) over the other (the one specified in the AWS credentials).

> Support both types would be nice but I'm not sure it is easy.

As far as I can tell it would require to find credentials of type `StandardCredentials` and then do a conditional on the type of implementation. Could be part of a different issue maybe ?

cyrille-leclerc replied:
> Could be part of a different issue maybe ?

It would definitively be a different Jira issue. I'm not sure that the Credentials API easily supports it at the moment.
